### PR TITLE
Add `Sequential` serialization test

### PR DESCRIPTION
**Pull request by niboshi**
Current test looks insufficient.

### CODE REVIEW

File: tests/chainer_tests/test_sequential.py [hunk 1]
```diff
@@ -581,4 +557,148 @@ def test_flatten(self):
...
+    # with an integer specified by `n`.
+
+    # `create_fn` is a test-specific function to create a target link.
+    # `create_fn` has one argument `nth`, which is either 0 or 1, each of
+    # which represents before or after serialization.
+    # `n` argument of `MyLink` is dependent on this `nth`, thus the link is
+    # initialized with different values depending on before/after
+    # serialization.
+
+    # If serialization dis not work correctly, the link would have a different
```
Comment (mitmul) [inline tests/chainer_tests/test_sequential.py]:
```suggestion
    # If serialization did not work correctly, the link would have a different
```

File: tests/chainer_tests/test_sequential.py [hunk 2]
```diff
@@ -581,4 +557,148 @@ def test_flatten(self):
         self.assertIs(flattened_s2[2], self.l3)
 
 
+class TestSequentialSerialization(unittest.TestCase):
+
+    # MyLink is a link that depends on a parameter whose value is filled
+    # with an integer specified by `n`.
+
+    # `create_fn` is a test-specific function to create a target link.
```
Comment (mitmul) [inline tests/chainer_tests/test_sequential.py]:
```suggestion
    # `fn_create` is a test-specific function to create a target link.
```

File: tests/chainer_tests/test_sequential.py [hunk 3]
```diff
@@ -581,4 +557,148 @@ def test_flatten(self):
...
+    # `create_fn` has one argument `nth`, which is either 0 or 1, each of
```
Comment (mitmul) [inline tests/chainer_tests/test_sequential.py]:
```suggestion
    # `fn_create` has one argument `nth`, which is either 0 or 1, each of
```

**Comment by mitmul** [comment]
@niboshi Could you resolve the conflicts?

**Comment by niboshi** [comment]
@mitmul 
Thanks. Rebased.

**Comment by mitmul** [comment]
Jenkins, test this please

**Comment by mitmul** [comment]
Jenkins, test this please

**Comment by mitmul** [comment]
ChainerCI, test this please

**Comment by mitmul** [comment]
Jenkins, test this please

**Comment by toslunar** [comment]
Could you resolve the conflicts?

**Comment by niboshi** [comment]
Resolved.

File: tests/chainer_tests/test_sequential.py [hunk 4]
```diff
@@ -607,4 +581,148 @@ def test_empty_sequential(self):
...
+
+    def my_func(self, n):
+        # defined as function-local function
+        def func_impl(x):
+            y, = self.MyFunc(n).apply((x,))
+            return y
+        return func_impl
+
+    def serialize(self, seq):
+        f = six.io.BytesIO()
```
Comment (toslunar) [inline tests/chainer_tests/test_sequential.py]:
```suggestion
        f = six.BytesIO()
```

File: tests/chainer_tests/test_sequential.py [hunk 5]
```diff
@@ -607,4 +581,148 @@ def test_empty_sequential(self):
...
+        chainer.serializers.save_npz(f, seq)
+        return f.getvalue()
+
+    def deserialize(self, seq, bytes):
+        f = six.io.BytesIO(bytes)
```
Comment (toslunar) [inline tests/chainer_tests/test_sequential.py]:
```suggestion
        f = six.BytesIO(bytes)
```

**Comment by toslunar** [comment]
FYI, I wrote a pull request to remove `six.io`: benjaminp/six#285.

**Comment by niboshi** [comment]
Fixed as suggested.

File: tests/chainer_tests/test_sequential.py [hunk 6]
```diff
@@ -607,4 +581,148 @@ def test_empty_sequential(self):
...
+    def deserialize(self, seq, bytes):
+        f = six.BytesIO(bytes)
+        chainer.serializers.load_npz(f, seq)
+
+    def check_serialization(self, fn_create):
+        x = numpy.array([[100, 100]], numpy.float32)
+        seq1 = fn_create(0)
+
+        y1 = seq1(x)
+        if len(seq1) == 0:
```
Comment (toslunar) [inline tests/chainer_tests/test_sequential.py]: Calling an empty `Sequential` is forbidden since #6304.

**Comment by mitmul** [comment]
Jenkins, test this please

**Comment by mitmul** [comment]
@niboshi Could you address this problem?
https://github.com/chainer/chainer/pull/6375#discussion_r270725817

**Comment by mitmul** [comment]
@niboshi Is just removing this condition fine? Is there a test case that uses an empty Sequential object? I'd like to resolve this problem:
https://github.com/chainer/chainer/pull/6375#discussion_r270725817

**Comment by niboshi** [comment]
@mitmul Sorry, I think I overlooked the comment. Fixed it.

File: tests/chainer_tests/test_sequential.py (hunk 6 above)
Comment (niboshi) [reply tests/chainer_tests/test_sequential.py]: Thank you for pointing out. Fixed.

**Comment by mitmul** [comment]
Thanks!

Jenkins, test this please

**Comment by mitmul** [comment]
Jenkins, test this please

**Comment by mitmul** [comment]
@niboshi `np.load` method is emitting an error in the Travis test. Could you check it and resolve the error?

**Pull request closed without merging**